PS-9148 feature: Dictionary caching for Masking Functions Component (8.0)

percona-ysorokin opened the pull request:
https://perconadev.atlassian.net/browse/PS-9148

dlenev commented on components/masking_functions/src/masking_functions/registration_routines.cpp:
```diff
@@ -960,41 +958,35 @@ class gen_blocklist_impl {
 
       if (!sresult) {
-        return {std::string{cs_term.get_buffer()}};
+        return cs_term_escaped;
```
So if term is not present in the dictionary our UDF returns its escaped version? And if it is present, we return some non-escaped random replacement term from the dictionary? This difference escaped - non-escaped sounds suspicious to me. Is this intentional change or a bug?

percona-ysorokin replied:
This layer has been significantly reworked:

1. Dictionary cache now always stores values in ubt8mb4 (no SQL escaping here).
2. If necessary, terms coming from UDF parameters are converted into utf8mb4 for interaction with dictionary cache.
3. SQL escaping happens only inside the `sql_builder` class.

dlenev commented on components/masking_functions/include/masking_functions/basic_sql_context_builder.hpp:
```diff
@@ -0,0 +1,54 @@
+
+namespace masking_functions {
+
+class basic_sql_context_builder {
```
It took me some time to figure out class hierarchy for basic_sql_context_builder and its descendants. Perhaps it would be easier if use more standard "abstract_sql_context_builder" for this class, and perhaps add comments describing this class as well as its descendants? What do you think?

percona-ysorokin replied:
`basic_sql_context_builder` renamed to `abstract_sql_context_builder`.
Added a comment describing class hierarchy and purpose of each class.

dlenev commented on components/masking_functions/src/masking_functions/registration_routines.cpp:
```diff
@@ -1025,20 +1017,53 @@ class gen_dictionary_impl {
+      throw std::invalid_argument{
+          "Wrong argument list: masking_dictionaries_flush()"};
+
+    ctx.mark_result_nullable(true);
```
I think it is worth to add a comment that will explain why this function is nullable/in which cases it can return NULL.

percona-ysorokin replied:
No longer nullable

dlenev commented on components/masking_functions/src/masking_functions/registration_routines.cpp:
```diff
@@ -960,41 +958,35 @@ class gen_blocklist_impl {
     const auto cs_term = make_charset_string_from_arg(ctx, 0);
-    const auto cs_dict_a = make_charset_string_from_arg(ctx, 1);
-    const auto cs_dict_b = make_charset_string_from_arg(ctx, 2);
+    const auto cs_term_escaped = escape_string(cs_term);
```
After a bit more thinking. IMO it is conceptually wrong to do escaping of dictionary and term names on this level (i.e. in UDF implementation). It is something which belongs to the level interacting with SQL. For example, if we will decide to change the implementation and start using prepared statements on the level below, such escaping on a high-level will become unnecessary and will only harm. So I'd like to propose move escaping back to query_builder.cpp layer.
What do you think?

percona-ysorokin replied:
SQL escaping on that level was definitely a wrong decision - this has been reworked properly now.

dlenev commented on components/masking_functions/include/masking_functions/bookshelf.hpp:
```diff
@@ -0,0 +1,58 @@
+  // returns empty std::string_view if no such dictionary exist
+  std::string_view get_random(
+      const std::string &dictionary_name) const noexcept;
+  bool remove(const std::string &dictionary_name) noexcept;
```
I think it is worth adding comment to mention what do return values for remove() and insert() mean.
For contains() the name is more or less self-explanatory, but for remove() and insert() it is not obvious, especially in MySQL context.

percona-ysorokin replied:
Done

dlenev commented on components/masking_functions/include/masking_functions/component_sys_variable_service_tuple.hpp:
```diff
@@ -0,0 +1,47 @@
+
+namespace masking_functions {
+
+// A set of MySQL query services required to perform system variable
```
Perhaps remove word "query" from this description?

percona-ysorokin replied:
Done

dlenev commented on components/masking_functions/include/masking_functions/default_sql_context_builder.hpp:
```diff
@@ -0,0 +1,41 @@
+
+namespace masking_functions {
+
+class default_sql_context_builder : public basic_sql_context_builder {
```
Perhaps it is worth to add a comment describing this class and its difference in comparison to static_sql_context_builder ?

percona-ysorokin replied:
Done

dlenev commented on components/masking_functions/include/masking_functions/dictionary.hpp:
```diff
@@ -0,0 +1,57 @@
+  bool contains(const std::string &term) const noexcept;
+  // returns empty std::string_view if the dictionary is empty
+  std::string_view get_random() const noexcept;
+  bool insert(const std::string &term);
```
Again IMO it would be nice to document meaning of return values for insert and remove() as it is not obvious.

percona-ysorokin replied:
Done

dlenev commented on components/masking_functions/include/masking_functions/bookshelf.hpp:
```diff
@@ -0,0 +1,58 @@
+
+namespace masking_functions {
+
+class bookshelf {
```
IMO it would be nice to add a comment describing what bookshelf class represents. So one won't have to look into its implementation to figure it out. Indeed bookshelf is place where books (in this case dictionaries) are stored, but still IMO it is worth to spell it out explicitly (e.g. we can say something like: "Bookshelf is a collection of objects representing dictionaries used for data-masking with dictionary name as an index").

percona-ysorokin replied:
Done

dlenev commented on components/masking_functions/include/masking_functions/dictionary.hpp:
```diff
@@ -0,0 +1,57 @@
+
+namespace masking_functions {
+
+class dictionary {
```
Again, in my opinion, since you are adding new class representing data-masking dictionary it would be nice to add short comment explaining what it is (e.g. collection of terms) and how it is used by data masking functionality on a high-level.

percona-ysorokin replied:
Done

dlenev commented on components/masking_functions/include/masking_functions/dictionary_flusher_thread.hpp:
```diff
@@ -0,0 +1,120 @@
+//   'deinit()' handlers as that may mess up with the 'THD' object of the
+//   connection that executes 'INSTALL COMPONENT' / 'UNINSTALL COMPONENT'.
+//   Therefore, 'INSTALL COMPONENT' cannot immediately reload dictionary
+//   cache. It can only spawn a background thread that will do this later.
```
Great comments! However, IMO, it would help if you have mentioned explicitly here that reload of dictionary will be done by this background thread OR by one of data masking UDFs trying to access it, whoever gets to the cache first.

In the end I have found this comment in another file, but having this information in this place would helped me a lot.

percona-ysorokin replied:
Done

dlenev commented on components/masking_functions/include/masking_functions/query_cache.hpp:
```diff
@@ -0,0 +1,60 @@
+
+namespace masking_functions {
+
+class query_cache {
```
Hmm... "Query Cache" has pretty specific meaning in MySQL context, so using this name here might create some confusion, in my opinion. And technically, you are not caching queries in it or their results, but rather the whole dictionary.

So perhaps it is a good idea to use some other name for this and  query_cache_core classes? dictionaries_cache perhaps? Or masking_dict_cache ? or something else...

What do you think?

dlenev replied:
After a bit more thinking... IMO it is definitely worth adding comment explaining relation between query_cache and query_cache_core, otherwise it is really not obvious why we have primitive singleton which stores pointer to query_cache, which stores pointer to query_cache_core, which stores pointer to bookshelf, where we cache things for real (too much indirection for my taste...).

It might help if we come up with a better names for these classes, after all, AFAICS query_cache class not only contains cache, but also sql_context_builder which controls the way UDFs and background thread work with data masking tables.

percona-ysorokin replied:
`query_cache` renamed to `term_cache`
`query_cache_core` renamed to `term_cache_core`
Added class descriptions.

dlenev commented on components/masking_functions/include/masking_functions/query_cache.hpp:
```diff
@@ -0,0 +1,60 @@
+  // returns a copy of the string to avoid race conditions
+  // an empty string is returned if the dictionary does not exist
+  std::string get_random(const std::string &dictionary_name) const;
+  bool remove(const std::string &dictionary_name);
```
As I have said earlier I think it is good idea to document return values from remove() and insert() methods, since they are non-obvious.

percona-ysorokin replied:
Done

dlenev commented on components/masking_functions/include/masking_functions/query_cache.hpp:
```diff
@@ -0,0 +1,60 @@
+ private:
+  query_cache_core_ptr core_;
+  basic_sql_context_builder_ptr sql_ctx_builder_;
+  query_builder_ptr sql_query_builder_;
```
As far as I can see we have only one instance of query_builder class in the component.
Maybe instead of passing pointer to it around as member function parameter, it makes sense to make it global singleton or simply static member in some class? What do you think?

percona-ysorokin replied:
Good point. Reworked as a global singleton - this simplified a lot of interfaces.

dlenev reviewed (changes requested):
Hello Yura!

Here are bit more comments/questions from me about the PR.

I am continuing my review meanwhile.

dlenev commented on mysql-test/suite/component_masking_functions/t/wait_for_component_uninstall.inc:
```diff
@@ -0,0 +1,35 @@
+# When flusher thread is enabled in masking functions component,
```
Such approach looks OK to me for case when explicit UNINSTALL COMPONENT is issued. But I wonder what happens if component deinit happens implicitly as result of shutdown. I.e. what will happen if server starts up, inits component and creates background thread, and then immediately tries to shutdown (e.g. due to some other component init failing).

Will the component deinit work correctly in this case?

percona-ysorokin replied:
From what I saw during debugging MySQL server during shutdown completely ignores the statuses of the component `deinit()` functions. So, there is a change that after our `deinit()` function completes (with error code), our background thread will still be alive. However, becase we registered our thread in the session manager, another part of the MySQL Server shutdown logic will iterate over remaining active sessions and will make several attempts of gracefully killing them (by setting `THD->killed` flag). One of the most recent flusher thread modifications made sure thsat we react properly to the `THD->killed`. So, this scenario is covered as well.

dlenev commented on mysql-test/suite/component_masking_functions/r/dictionary_operations.result:
```diff
@@ -21,59 +28,71 @@ ERROR HY000: Can't initialize function 'gen_blocklist'; Wrong argument list: gen
 SELECT gen_blocklist('Berlin', 'de_cities', NULL);
 ERROR HY000: gen_blocklist<string> UDF failed; argument 3 cannot be null
+#
 # checking the case when there is no mysql.masking_dictionaries table
```
I think, that after changes in this PR you no longer test case here case when mysql.masking_dictionaries is missing, but instead case when mysql_session@localhost user has insufficient privileges on it... Perhaps the test needs to be extended to test both these scenarios, or at least comment should be adjusted?

percona-ysorokin replied:
See previous response
https://github.com/percona/percona-server/pull/5520?timeline_per_page=10#discussion_r1913972030

percona-ysorokin replied:
MTR test case extended with both scenarios.

dlenev commented on mysql-test/suite/component_masking_functions/t/flusher_thread_immediate_restart.test:
```diff
@@ -0,0 +1,22 @@
+INSERT INTO mysql.masking_dictionaries VALUES('dict1', 'term1');
+
+
+INSTALL COMPONENT 'file://component_masking_functions';
```
OK. I guess this test will occasionally trigger scenario with immediate shutdown and component deinit failure I have mentioned... But I wonder how often it will happen... Perhaps we can make situation more reproducible by adding debug sleep at various stages of flusher thread initialization? What do you think?

percona-ysorokin replied:
See previous response

dlenev commented on mysql-test/suite/component_masking_functions/t/flusher_thread_suspend_resume.test:
```diff
@@ -0,0 +1,69 @@
+--let $assert_text = The very first call to gen_dictionary("dict2") should not trigger inplace cache reload and return NULL
+--source include/assert.inc
+
+SET debug_sync = 'now SIGNAL masking_functions_before_cache_reload_signal WAIT_FOR masking_functions_after_cache_reload_signal';
```
A good style is to do debug_sync = 'RESET' at the end of part of the test case using debug sync facility. Yes, in your case problems are unlikely due to presence of .opt file for test case, but IMO it is good idea to stick to good style anyway.

percona-ysorokin replied:
Done

dlenev commented on mysql-test/suite/component_masking_functions/t/flusher_thread_connection_reuse.test:
```diff
@@ -0,0 +1,40 @@
+
+--source wait_for_component_uninstall.inc
+
+SET GLOBAL debug = '-d,enable_masking_functions_flusher_create_sync,enable_masking_functions_flush_thread_sync,enable_masking_functions_flush_thread_double_pass';
```
I think it is a good idea to reset debug_sync facility at the end of the test as well.

percona-ysorokin replied:
Done

dlenev commented on mysql-test/suite/component_masking_functions/t/rpl_dictionaries_flush_interval.test:
```diff
@@ -0,0 +1,138 @@
+# Cleanup
+--source include/rpl_connection_slave.inc
+REVOKE MASKING_DICTIONARIES_ADMIN ON *.* FROM root@localhost;
+SET GLOBAL debug ='-d,enable_masking_functions_flush_thread_sync';
```
Ditto here. I think it is a good style to reset debug_sync facility at the end of the test as well for both master and slave (though we are not likely to see any problems for this specific test due to such resetting missing).

percona-ysorokin replied:
Done

dlenev reviewed (changes requested):
Hello Yura!

I have a bit more comments about your PR.

percona-ysorokin commented:
This PR is suspended. All its comment will be addressed in the new one
https://github.com/percona/percona-server/pull/5535
(rebased on 8.0.40, squashed commits, fixed clang-tidy warnings for `component_masking_functions` code).
